### ISSUE TITLE
Orchestrator fails randomly with "Non-Deterministic workflow detected" message

### ISSUE THREAD

**Issue by 10gmp** [Azure/azure-functions-durable-extension#583]
We are having an orchestrator that randomly fails with "Non-Deterministic workflow detected" when calling "CallActivityAsync". We have tested the same orchestrator with the same data and sometimes it fails but sometimes doesn't. The process runs for multiple hours and fails at different times.

- Durable Functions extension version: 1.7.1
- Function App version: 1.0
- Programming language used: C# (Compiled/Visual Studio)
- Timeframe issue observed: 2019-01-22T16:46:11.023Z
- Region: South Central
- Orchestration instance: afb6fceb6b8844a683ef54d2d70d76a7

**Comment by cgillum** [comment]
Thanks for these details. In most cases, a non-deterministic orchestration is caused by incorrectly written orchestrator function code. Can you share a sample of what your orchestration code looks like to help rule that out?

**Comment by 10gmp** [comment]
Here we query a list o batches and depending on the stage and status we trigger the LogResults activity. After that we add the Task to a lists of LogResult running tasks.
```c#
foreach (Entities.Batch batch in batches.Where(b => b.StageCode == Entities.BatchStage.Log && b.StatusCode == Entities.BatchStatus.Pending))
{
    Task<(Dictionary<string, int> faults, double duration)> task = context.CallActivityAsync<(Dictionary<string, int>, double)>("LogResults", (settings, batch));
    batch.StatusCode = Entities.BatchStatus.Running;
    batch.Thread = task.Id;
    batch.Duration = TimeSpan.FromSeconds(0);
    resulTasks.Add(task);
    LogBatch(batch);
}
```
Here we wait for any Task in three running task lists we have.
```c#
if (instance.StatusCode != Entities.InstanceStatus.Completed &&  instance.StatusCode != Entities.InstanceStatus.Stopped &&  instance.StatusCode != Entities.InstanceStatus.Failed)
{
    List<Task> waitFor = new List<Task>();
    waitFor.AddRange(extractTasks);
    waitFor.AddRange(loadTasks);
    waitFor.AddRange(resulTasks);
    if (waitFor.Count > 0)
        await Task.WhenAny(waitFor);
}
```
Here we go thru the running task list looking for completed and process them.
```c#
var completedResult = resulTasks.Where(t => t.IsCompleted);
foreach (var task in completedResult)
{
    Entities.Batch batch = batches.First(b => b.Thread == task.Id);
    if (task.IsFaulted || task.IsCanceled) FailActivity(task, batch);
    else
    {
        batch.StatusCode = Entities.BatchStatus.Finished;
        (Dictionary<string, int> faults, double duration) = task.Result;
        batch.Duration = TimeSpan.FromSeconds(duration);
        batch.FaultTypes = faults;
        instance.LogTime += batch.Duration;
        foreach (var fault in faults)
        {
            if (instance.FaultTypes.Keys.Contains(fault.Key)) instance.FaultTypes[fault.Key] += fault.Value;
            else instance.FaultTypes[fault.Key] = fault.Value;
        }
        LogBatch(batch);
    }
}
```
This is the LogResults activity signature
```c#
[FunctionName("LogResults")]
public static (Dictionary<string, int>, double) Run([ActivityTrigger] DurableActivityContext inputs, ILogger log)
{
    ...
}
```

**Comment by cgillum** [comment]
First, can you confirm that `extractTasks`, `loadTasks`, and `resulTasks` are all valid "durable" tasks?

Secondly, this code seems very suspicious to me:
```cs
Entities.Batch batch = batches.First(b => b.Thread == task.Id);
```
My worry is that task and/or thread IDs may not be deterministic. Looking at the [documentation for Task.Id](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.id?view=netframework-4.7.2), it seems reasonable to me that this property is not safe to use in an orchestrator function:

> Task IDs are assigned on-demand and do not necessarily represent the order in which task instances are created. Note that although collisions are very rare, task identifiers are not guaranteed to be unique.

I think it's very likely that this line of code is causing your problem. Try modifying your code to not rely on runtime-generated task IDs and see if that resolves the issue.

**Comment by 10gmp** [comment]
Yes, I can confirm extractTasks, loadTasks and resultTasks are all valid durable tasks.
I understand what you say about Task Ids, but from what I read, the worst case scenario will be not finding any batch and getting a "Sequence contains no elements" exception when calling the "First" method.. I don't understand why this is causing a Non-Deterministic workflow.

**Comment by 10gmp** [comment]
Is there any thing else that could be the root cause of this issue?

**Comment by cgillum** [comment]
> I understand what you say about Task Ids, but from what I read, the worst case scenario will be not finding any batch and getting a "Sequence contains no elements" exception when calling the "First" method.. I don't understand why this is causing a Non-Deterministic workflow.

I was thinking it could happen if `.First(...)` simply returned a different task on different replays. But even if it failed with "Sequence contains no elements", that could potentially result in the non-deterministic workflow error if it is encountered on a replay when the original execution succeeded.

That said, I looked at your instance and I didn't see any of those errors. Rather, I saw this:

```
System.ArgumentException: The function 'LogErrors' doesn't exist, is disabled, or is not an activity function. Additional info: No activity functions are currently registered!
   at Mix.Functions.Engine.<Run>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Executors.VoidTaskMethodInvoker`2.<InvokeAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.<InvokeAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Executors.TriggeredFunctionExecutor`1.<>c__DisplayClass6_1.<<TryExecuteAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskOrchestrationShim.<Execute>d__8.MoveNext()
```
The timestamp is `2019-01-22 16:46:11.8111151`, which matches almost exactly with the timestamp you provided. I wonder if this might actually be the cause of the non-deterministic workflow detected.

I looked further, and I also noticed this in the general purpose Azure Functions logs a few minutes earlier: 
```
Failed to load type 'Mix.Functions.LogError' from 'D:\home\site\wwwroot\bin\Engine.dll'
```
```
The following 1 functions are in error:
LogError: The function type name 'Mix.Functions.LogError' is invalid.
```
Are you aware of any issues with your `LogError` function?

**Comment by 10gmp** [comment]
I'm not aware of any error on that Function, in fact it was fired about 900 times during that orchestrator lifetime.

**Comment by 10gmp** [comment]
If it helps, this is the way we are calling the LogErros Function.
```c#
if (instance.FaultTypes.Count > 0)
{
    await context.CallActivityAsync("LogErrors", (instance.FaultTypes, instance.Job, instance.InstanceSequence, Entities.ErrorType.Fault));
}
if (instance.FailureTypes.Count > 0)
{
    await context.CallActivityAsync("LogErrors", (instance.FailureTypes, instance.Job, instance.InstanceSequence, Entities.ErrorType.Failure));
    instance.Failures = batches.Count(b => b.StatusCode == Entities.BatchStatus.Failed);
}
```
And this is the signature of the Functino.
```c#
[FunctionName("LogErrors")]
public static void Run([ActivityTrigger] DurableActivityContext inputs, ILogger log)
{
    ...
}
```

**Comment by 10gmp** [comment]
It has happened again on orchestrator: 794d3b7e380f4265a3404886840a46a7, with error: "Non-Deterministic workflow detected: TaskScheduledEvent: 3994 TaskScheduled LogErrors". You can see the error on History table but not on the Instance table, the instance is still in Running status.

**Comment by 10gmp** [comment]
We keep getting the same issue on different functions fired by our orchestrator. @cgillum is there anything you can do to help us diagnose or fix this problem.

**Comment by farzanDE** [comment]
I'm also facing the same issue. I have a long running orchestrator (6-8 hours) that runs daily. In some cases, it runs to completion but occasionally fails (at random points, not at a specific activity function) with the "Non-deterministic workflow" message. If I re-execute it without any changes, it eventually goes to completion. Recently the frequency of failures has increased greatly so it is becoming impractical to keep re-executing until it succeeds and I'm looking for a permanent solution. Any suggestions?

**Comment by GlennDOrr** [comment]
I'm getting the same behavior, what I've found in my case is that there is an exception being fired in the activity and error is being reported as Non-Deterministic workflow detected: TaskScheduledEvent: ...

**Comment by ConnorMcMahon** [comment]
@10gmp,

Sorry this issue fell by the wayside. I know it has been over a year, but if you are still encountering it, give me an orchestration instance id and we can investigate. Overall, given what @cgillum found, it looks like the Functions runtime/platform was having an intermittent issue finding one of your activity functions, which led to the non-determinism exception. Hopefully it has been resolved by now due to the improvements made to the Azure Functions service.

@farzanDE and @GlennDOrr, in general, non-determinism errors have unique causes, and frequently due to errors inside of orchestration code. The non-determinism errors you have are unlikely to be the same root cause as the original issue. If you are encountering them still, I recommend opening a new ticket, with both an orchestration instance id for us to look at, along with a timestamp, region, and if at all possible segments of your orchestration code.

**Comment by cgillum** [comment]
I would also add that we've also made several fixes in the durable extension around Function registration which may have resolved issues related to function discovery (which could explain issues with functions not being found). It's quite possible that the original issue has been fixed.

A couple other things I'd recommend in general regarding non-determinism errors:

1. Take a look at our [Durable Functions Analyzer](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers). It can help detect common coding issues that sometimes get overlooked.
2. Take a look at your [Versioning documentation](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-versioning). If you are deploying code changes that could impact your in-flight, long-running orchestrations, then that could also produce non-determinism errors.
3. Make sure you're using the latest version of the extension. Each release contains a good number of fixes. If you continue to run into issues, do let us know what version of the extension you're using.
